examples/BytePairEncoder.hs

```diff
@@ -0,0 +1,219 @@
+-- To run this program:
+--
+-- cabal run --flag fusion-plugin BytePairEncoder test-data.txt
+--
+module BytePairEncoder (main) where
+
+import Control.Monad.IO.Class (MonadIO)
+import Data.Function ((&))
+import Data.List (maximumBy)
+import qualified Data.Map as M
+import Data.Maybe (fromJust, fromMaybe)
+import Data.Ord (comparing)
+import qualified Data.Vector as V
+import GHC.Word (Word8)
+import qualified Streamly.Data.Array as Array
+import qualified Streamly.Data.Fold as Fold
+import Streamly.Data.Stream (Stream)
+import qualified Streamly.Data.Stream as Stream
+import qualified Streamly.FileSystem.File as File
+import Streamly.Internal.Data.Pipe (Pipe (..), Step (..))
+import Streamly.Internal.Data.Stream (pipe)
+import System.Environment (getArgs)
+
+-------------------------------------------------------------------------------
+-- Byte indexing and text representation
+-------------------------------------------------------------------------------
+
+-- Stores byte-sequence-to-index mapping and index-to-text mapping
+data ByteMappings = ByteMappings
+  { byteToIndex :: !(M.Map Word8 Int), -- Maps bytes to unique indices
+    seqToIndex :: !(M.Map (V.Vector Word8) Int), -- Maps sequences of bytes to unique indices
```

Comment:
I'm curious to see how a cuckoo hash table might behave in this case.
We can try using it and checking performance.

Reply:
You can use an unboxed Array instead of a `Vector Word8`
`Array Word8` basically.